llvm/include/llvm/CAS/ActionCache.h

```diff
@@ -0,0 +1,98 @@
+//===- llvm/CAS/ActionCache.h -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_CASACTIONCACHE_H
+#define LLVM_CAS_CASACTIONCACHE_H
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/CASReference.h"
+#include "llvm/Support/Error.h"
+
+namespace llvm::cas {
+
+class CASDB;
+class CASID;
+class ObjectProxy;
+
+/// A key for caching an operation.
+/// It is implemented as a bag of bytes and provides a convenient constructor
+/// for CAS types.
+class CacheKey {
+public:
+  StringRef getKey() const { return Key; }
+
+  // TODO: Support CacheKey other than a CASID but rather any array of bytes.
+  // To do that, ActionCache need to be able to rehash the key into the index,
+  // which then `getOrCompute` method can be used to avoid multiple calls to
+  // has function.
+  CacheKey(const CASID &ID);
+  CacheKey(const ObjectProxy &Proxy);
+  CacheKey(const CASDB &CAS, const ObjectRef &Ref);
+
+private:
+  std::string Key;
+};
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (collision is an error).
+class ActionCache {
+  virtual void anchor();
+
+public:
+  /// Get a previously computed result for \p ActionKey.
+  Expected<Optional<ObjectRef>> get(const CacheKey &ActionKey) const {
+    return getImpl(arrayRefFromStringRef(ActionKey.getKey()));
+  }
+
+  /// Cache \p Result for the \p ActionKey computation.
+  Error put(const CacheKey &ActionKey, const ObjectRef &Result) {
+    return putImpl(arrayRefFromStringRef(ActionKey.getKey()), Result);
+  }
+
+  /// Cache \p Result using \p ObjectRef as a key.
+  Error put(const ObjectRef &RefKey, const ObjectRef &Result) {
+    CacheKey ActionKey(CAS, RefKey);
+    return putImpl(arrayRefFromStringRef(ActionKey.getKey()), Result);
+  }
+
+  /// Get or compute a result for \p ActionKey. Equivalent to calling \a get()
+  /// (followed by \a compute() and \a put() on failure).
+  Expected<ObjectRef> getOrCompute(const CacheKey &ActionKey,
+                                   function_ref<Expected<ObjectRef>()> Compute);
+
+  virtual ~ActionCache() = default;
+
+protected:
+  virtual Expected<Optional<ObjectRef>>
+  getImpl(ArrayRef<uint8_t> ResolvedKey) const = 0;
+  virtual Error putImpl(ArrayRef<uint8_t> ResolvedKey,
+                        const ObjectRef &Result) = 0;
+
+  ActionCache(CASDB &CAS) : CAS(CAS) {}
+
+  CASDB &getCAS() const { return CAS; }
```

Comment:
If we made this public, we could potentially avoid needing to pass both a CAS and an ActionCache everywhere.  It seems like to use the cache you are already tied to using a specific CAS since otherwise the ObjectRef won't make sense.  WDYT?

Reply:
That doesn't really work unless we make ActionCache owns CAS, otherwise this is going to open to all kinds of life time issues.